sycl/plugins/unified_runtime/ur/adapters/level_zero/usm.hpp

```diff
@@ -40,75 +40,100 @@ class UsmAllocationException {
     return UMF_RESULT_ERROR_NOT_SUPPORTED;
   };
-  umf_result_t purge_force(void *, size_t) {
+  virtual umf_result_t get_min_page_size(void *, size_t *) {
     return UMF_RESULT_ERROR_NOT_SUPPORTED;
   };
+  virtual umf_result_t get_recommended_page_size(size_t, size_t *) {
+    return UMF_RESULT_ERROR_NOT_SUPPORTED;
+  };
+  virtual umf_result_t purge_lazy(void *, size_t) {
+    return UMF_RESULT_ERROR_NOT_SUPPORTED;
+  };
+  virtual umf_result_t purge_force(void *, size_t) {
+    return UMF_RESULT_ERROR_NOT_SUPPORTED;
+  };
+  virtual const char *get_name() { return ""; };
+  virtual ~USMMemoryProviderBase() = default;
+};
+
+// Implements USM memory provider interface for L0 RT USM memory allocations.
+class L0MemoryProvider : public USMMemoryProviderBase {
+private:
+  // Min page size query function for L0MemoryProvider.
+  std::pair<umf_result_t, size_t> GetL0MinPageSize();
+  size_t MinPageSize = 0;
+  bool MinPageSizeCached = false;
+
+public:
+  umf_result_t initialize(ur_context_handle_t Ctx,
+                          ur_device_handle_t Dev) override;
+  umf_result_t alloc(size_t Size, size_t Align, void **Ptr) override;
+  umf_result_t free(void *Ptr, size_t Size) override;
+  umf_result_t get_min_page_size(void *, size_t *) override;
   // TODO: Different name for each provider (Host/Shared/SharedRO/Device)
-  const char *get_name() { return "USM"; };
-  virtual ~USMMemoryProvider() = default;
+  const char *get_name() override { return "L0"; };
 };
 
 // Allocation routines for shared memory type
-class USMSharedMemoryProvider final : public USMMemoryProvider {
+class L0SharedMemoryProvider final : public L0MemoryProvider {
 protected:
   ur_result_t allocateImpl(void **ResultPtr, size_t Size,
                            uint32_t Alignment) override;
 };
 
 // Allocation routines for shared memory type that is only modified from host.
-class USMSharedReadOnlyMemoryProvider final : public USMMemoryProvider {
+class L0SharedReadOnlyMemoryProvider final : public L0MemoryProvider {
 protected:
   ur_result_t allocateImpl(void **ResultPtr, size_t Size,
                            uint32_t Alignment) override;
 };
 
 // Allocation routines for device memory type
-class USMDeviceMemoryProvider final : public USMMemoryProvider {
+class L0DeviceMemoryProvider final : public L0MemoryProvider {
 protected:
   ur_result_t allocateImpl(void **ResultPtr, size_t Size,
                            uint32_t Alignment) override;
 };
 
 // Allocation routines for host memory type
-class USMHostMemoryProvider final : public USMMemoryProvider {
+class L0HostMemoryProvider final : public L0MemoryProvider {
 protected:
   ur_result_t allocateImpl(void **ResultPtr, size_t Size,
                            uint32_t Alignment) override;
 };
 
-// Simple proxy for UMF allocations. It is used for the UMF tracking
+// Simple proxy for L0 allocations. It is used for the UMF tracking
 // capabilities.
-class USMProxyPool {
+class L0ProxyPool {
```

Comment:
Is it somehow specific to L0? I guess we could use it for other adapters as well?

Reply:
My bad. I've been using it only in the context of L0.